Add `<when>` to help select the right `<match>` (#558)
unicode-org/message-format-wg: spec/registry.md

If multiple `<override>` elements would match the current locale,
only the first one is used.

### Variant Key Matches

> Matching-function signatures can include `<matches>` and `<when>` elements

Author: Do you think we should (separately from this PR) also be renaming the `<matchSignature>` that's the actual element within which these may be included?

> Matching-function signatures can include `<matches>` and `<when>` elements

Reviewer: I don't think we need to. But in this very same document we already call the "matching-function" a "selector". And we call it that in syntax.md, etc.

defining the variant keys matched by the selector.

> Each `<matches>` MAY contain either one or more `<match>` elements, or an `href` attribute.

Author: That would preclude the possibility of having no `href` and no `<match>` elements. That could happen if there's an `href` to an XML that's generated from other data, and in some conditions that data might filter its list of `<match>` to none. As we do have support for not finding any `<match>` for the current locale, we can support that situation with no problems in the processing. Why then should we consider it an error?

> Each `<matches>` MAY contain either one or more `<match>` elements, or an `href` attribute.

Reviewer: Fine, but your phrasing (if that's the intent) is way too subtle. If you intend to allow zero match elements and also no href, you should explicitly state it. In that case...

```suggestion
Each `<matches>` contains either a (possibly empty) set of `<match>` elements or an `href` attribute.
```

If an `href` attribute is set, its URL value MUST resolve to an XML document

> with a root `<matches>` element with no `href` attribute,

Author: Because we've no reason to, and this way we can rely on the external matches XML to resolve all of its dependencies completely.

> with a root `<matches>` element with no `href` attribute,

Reviewer: So there are no cases in which part of a `<matches>` tree in the external XML refers to (say) CLDR data? Once you've implemented resolving an external file, it's just a question of recursion, no?

> with a root `<matches>` element with no `href` attribute,

Author: > So there are no cases in which part of a `<matches>` tree in the external XML refers to (say) CLDR data?

Correct. A `<matches>` can only contain `<match>` elements, which may only have `locales` and `values` attributes, and are otherwise empty. So there's no space for recursion.

which will then replace the current `<matches>` element for all later processing.
If `<matches>` contains any child elements, its `href` attribute is ignored.
Otherwise, if `<matches>` contains an `href` attribute, its `validationRule` attribute is ignored.

The `<match>` element whose `locales` best matches the current locale
using resource item [lookup](https://unicode.org/reports/tr35/#Lookup) from LDML is used.
An element with no `locales` attribute is the default
(and is considered equivalent to the `root` locale).

> As the available keys may depend on option values,

Author: The available keys don't _always_ depend on option values, though.
```suggestion
In situations where the available keys depend on option values,
```

`<when>` elements can be used to select an appropriate `<matches>` element for selection.
If the resolved or default value of a selector option
corresponding to the `<when>` `option` attribute
is included in its list of `values`,
its contents are considered before any and all later `<when>` and `<matches>` elements.
If a `<matches>` element within a `<when>` has a `<match>` for the current locale,

> later `<matches>` elements outside that `<when>` are not considered.

Reviewer: This is really hard to grok.

I think maybe this should go sooner, since `<when>` (when present) is a parent to `<matches>`. Something like:

> The results of a selector depend on the the value of options.
> If a `<when>` element is present that matches a selector `option`
> then only the elements contained by the `<when>` are used to evaluate
> the selector.

Question: what about nesting of `when`?

Author: The nesting is handled perhaps a bit too implicitly by the `(when*,matches?)` ordering, and the use of the term "contents" referring to both in the passage above.

The language you propose is a bit problematic, because
1. It makes it possible for two different `<when>` elements to both qualify for being "used to evaluate
the selector."
2. It presumes completeness of inner `<matches>`. For example, with the plural integer selector that I've used as an example, the `<matches>` for `maximumFractionDigits=0` might not include anything for `en` because English doesn't make a distinction on that axis. Then we need to look at the plural selection `<matches>`, which would include an `en` entry.

But your overall concern about the language here being hard to grok is on point, and it should be revised.

Reviewer: > If a `<matches>` element within a `<when>` has a `<match>` for the current locale, later `<matches>` elements outside that `<when>` are not considered.

Is this correct? Maybe the missing piece is:

```suggestion
If a `<matches>` element within a matching `<when>` has a `<match>` for the current locale,
later `<matches>` outside that `<when>` are not considered.
```

That is, we only want to consider this if the `when` elements options match, not for irrelevant `when` elements?

(Otherwise a good improvement)

Author: Yeah, that does improve things. Edited your suggestion to not drop the preceding 5 lines and just rephrase the last sentence.

```

Furthermore,
`:number`'s `<matchSignature>` contains multiple `<matches>` and `<when>` elements
which allow the validation of variant keys.

> - `<when option="select" values="plural"><matches><match locales="en" values="one other" ... />`
>   can be used in locales like `en` and `en-GB` if the selection type is known to be plural
>   to validate that only `one`, `other` or numeric keys are used for variants.

Reviewer: I think we might want to make these into examples. And I think we might want to avoid "validate that only". Perhaps:

```suggestion
> For example,
> `<when option="select" values="plural"><matches><match locales="en" values="one other" ... />`
> could be used when validating translations for locales such as `en` and `en-GB`
> to check that variant keys `one` and `other` have been provided
> (in addition to any numeric keys).
```
